feat: git status in telescope preview

gusjengis opened the pull request:
# Git Status Previews!
![image](https://github.com/user-attachments/assets/04ea4a9b-ede9-4e6d-a741-e9cb1fb17657)
This image pretty much shows off all of the features, I abused one of my repos so you could see all of the pretty colors.

List of changes:
- Added option to show hidden files in preview
- Added option to show git status info in preview
- Added header to preview. This includes the project title, current branch, and number of commits ahead and behind the origin
- Added status indicators to files, A, M, and D, for added, modified, and deleted
- Added lots of highlighting, should be compatible with any colorscheme
- Added a caching system for project previews, the picker was beginning to stutter from running git commands, still very usable without caching, but once the caching is done all hitches are eliminated
- Also added some debouncing to enable smooth key_repeat scrolling even without completed preview caching
- Now running git fetch asynchronously to get the current number of commits ahead and behind, you can find some comments that go more in depth on this
- Now running commands to enable maintenance and fsmonitor on all repos, this speeds up fetch and status on large repos and is harmless on small ones

##  Colorschemes
I think this looks best in my colorscheme of choice, github_dark_default, but I made an effort to pull colors from the current colorscheme, so it should work with anything. This seems to play nice with all of the themes installed on my system, but not all themes define the same highlight groups, so it's possible there are some out there that would have contrast issues.
#### default
![image](https://github.com/user-attachments/assets/feb57236-74fa-49e1-a887-fa7bb4e532df)
#### tokyonight-night
![image](https://github.com/user-attachments/assets/9ae69d1f-34fa-4fc1-b8d9-4d4ba852a668)
#### catppuccin-mocha
![image](https://github.com/user-attachments/assets/0a06da83-9e42-4d83-a3f9-115dbfba1038)
#### onedark (warmer)
![image](https://github.com/user-attachments/assets/15befd1b-4ac5-4dd1-94d3-636960642062)

This even works with hideous light themes

#### github-light-default
![image](https://github.com/user-attachments/assets/9659bbd8-b080-4ef5-81a4-f12404fdb1d0)
#### peachpuff
![image](https://github.com/user-attachments/assets/3705ae42-d5fe-4c31-b235-7dcc0816bd1b)


## Non-Repos
![image](https://github.com/user-attachments/assets/13011e5b-5291-43e5-ae1c-a779124c9a8c)
Project folders that aren't repos just appear as any project would with git_status set to false.

gusjengis commented:
I flirted with including more information, stuff like the current commit name and date of the last local commit or change would be useful. The commit name can be arbitrarily long though, which can ruin the appearance of the header. There are also highlighting issues with adding more info. The more colored backgrounds I add the harder it becomes to pick out contrasting colors across all colorschemes. Maybe the commit hash would be okay, it should always be the same size, but that would be meaningless to me. 

A last-modified date would be nice, or maybe a duration since last modification, I think I'd prefer that. 

I think that could be implemented by either checking the last commit time, or, if there are uncommitted changes, by taking the max of the vim.fn.getftime of all files listed by git status --porcelain. If the project is not a repo, then you could fall back to the max vim.fn.getftime of all top level files. Maybe I'll find time to implement this later, but this has sidetracked me long enough for now.

coffebar commented:
Thanks for your effort, looks good.

1. We need to update mentions of preview option in the installation instructions. For example [here](https://github.com/gusjengis/neovim-project/blob/6ee26bce0cf190c114e91cb5833be97dd5e4b16c/README.md?plain=1#L44) I'm okay if you will remove preview related config from all install instructions or update it to the same way as it is in full config.
2. `git config core.fsmonitor` - changing repo config in any way without asking user seems too much for me. We can't do this, even if it seems not harmful (when you have enough free resources and OS limits to run some additional background daemons with file watchers). Let's just remove this line, ok? 
 3. I could use some of your screenshots for README page, but github resizes them, so it's hard to see something without zooming. You can try to make your window size smaller (to reduce free space) and also maybe crop it more. But don't hurry with this, maybe we will change something in UI until this PR will be merged. I didn't test yet

gusjengis commented:
Okay, removed git config commands, the benefit was purely theoretical anyway, I don't have any large repos.

Also removed mentions of the preview setting from the readme, except for in the default config section.

I look forward to seeing a screenshot of it running on your setup, you seemed to have a theme I don't recognize, I'm curious how it will look.

coffebar commented:
I'm using [onedark warm](https://github.com/navarasu/onedark.nvim) colorscheme, but i changed some of colors to match [arc-dark](https://github.com/horst3180/arc-theme) GTK theme.

![image](https://github.com/user-attachments/assets/5a7cfc0b-6f0f-47e4-91fd-d98ec93413e7)
 

1. doc txt file should be updated as well https://github.com/gusjengis/neovim-project/blob/ce48d07ac6654bfde76ee1b2ba886f2a993fb367/doc/neovim-project.txt#L61
2. I noticed the global directory change in preview - `vim.fn.chdir.` This is a problem because the plugin itself has auto-commands for this event to switch projects, and other plugins can react to it, strange bugs with LSP servers and other things can occur. Can you try to get around it without chdir globally? 

```lua
--- my first thought in this direction
vim.fn.jobstart(..., { cwd = project_path })
```

When I scroll project list, sometimes I see this message, that tells me things went wrong (because of chdir)
![image](https://github.com/user-attachments/assets/cd40d79f-cde6-4434-8af4-345e63137c66)

gusjengis commented:
That's what that is! I've been seeing that notification, but I hadn't really looked closely at.

Should be fixed, docs are updated and I swapped chdir for proper parameters in the git commands.

Do you want more zoomed in screenshots? I assumed you'd add a screenshot from your own setup to keep things consistent.

coffebar commented on lua/neovim-project/preview.lua:
```diff
@@ -74,7 +74,7 @@ M.init = function()
   })
 
-  -- Set up an autocmd to clear caches periodically
+  -- Set up an autocmd to clear currect project cache when opening the picker, this ensures that recent changes are visible
```
typo in the comment "currect"

coffebar commented:
> Do you want more zoomed in screenshots? I assumed you'd add a screenshot from your own setup to keep things consistent.

As you wish

coffebar commented on doc/neovim-project.txt:
```diff
@@ -194,7 +194,11 @@ DEFAULT OPTIONS: ~
       picker = {
         type = "telescope", -- or "fzf-lua"
-        preview = true, -- show directory structure preview in telescope
+	preview = {
```
here something wrong with spaces

gusjengis replied:
?

coffebar replied:
![image](https://github.com/user-attachments/assets/22bc3b4f-e23c-4515-8dd8-5fe29a3b6b45)

coffebar replied:
Code is not aligned

gusjengis replied:
Looked normal on my screen, but I noticed forgot to include the preview config in the docs for all package managers, should be fine now.

coffebar replied:
It's not on my. The problem is this file has spaces for indentation, but you are using tabs. Tabs looks differently depending on config, so you don't see it.

gusjengis replied:
I wondered why people argue about tabs vs spaces, makes sense now. Should be fixed.

coffebar commented:
If user had config option `picker.preview = false` to disable preview, not it will not work. Have to replace with `picker.preview.enabled = false` to work
